### PR TITLE
Error out for const scalar return type in reduction

### LINKED ISSUE #2090: parallel_reduce with const return value reference doesn't fail like it should
The following snippet compiles, runs, and prints `0`:

```c++
    const int value = 0;

    Kokkos::parallel_reduce(42,
      KOKKOS_LAMBDA (int i, int& result) {
        result += 1;
      },
      value
    );

    Kokkos::fence();
    std::cout << value << std::endl;
```

This feels like a bug...

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #2090. This pull request also makes sure that we are calling the intended overload for `parallel_reduce`, i.e., the return value is only copied if the return type is a `View` or a reducer.

### CODE REVIEW

**Comment by masterleinad** [comment]
The linked example now gives
```
In file included from /tmp/kokkos-alt/core/src/Kokkos_Parallel.hpp:180:0,
                 from /tmp/kokkos-alt/core/src/Kokkos_OpenMP.hpp:62,
                 from /tmp/kokkos-alt/core/src/decl/Kokkos_Declare_OPENMP.hpp:49,
                 from /tmp/kokkos-alt/example/build_cmake_in_tree/build/kokkos/KokkosCore_Config_DeclareBackend.hpp:47,
                 from /tmp/kokkos-alt/core/src/Kokkos_Core.hpp:56,
                 from /tmp/kokkos-alt/example/build_cmake_in_tree/cmake_example.cpp:1:
/tmp/kokkos-alt/core/src/Kokkos_Parallel_Reduce.hpp: In instantiation of ‘std::enable_if_t<(!(Kokkos::is_view<Arg>::value || Kokkos::is_reducer<ReturnType>::value))> Kokkos::parallel_reduce(const size_t&, const FunctorType&, ReturnType&) [with FunctorType = main()::<lambda(int, int&)>; ReturnType = const int; std::enable_if_t<(!(Kokkos::is_view<Arg>::value || Kokkos::is_reducer<ReturnType>::value))> = void; size_t = long unsigned int]’:
/tmp/kokkos-alt/example/build_cmake_in_tree/cmake_example.cpp:8:68:   required from here
/tmp/kokkos-alt/core/src/Kokkos_Parallel_Reduce.hpp:1778:3: error: static assertion failed: A const return type is only allowed for a View or reducer return type!
   static_assert(
   ^~~~~~~~~~~~~
/tmp/kokkos-alt/core/src/Kokkos_Parallel_Reduce.hpp:1588:22: error: ‘static void Kokkos::Impl::ParallelReduceAdaptor<PolicyType, FunctorType, ReturnType>::execute(const string&, const PolicyType&, const FunctorType&, ReturnType&) [with PolicyType = Kokkos::RangePolicy<Kokkos::OpenMP>; FunctorType = main()::<lambda(int, int&)>; ReturnType = const int; std::__cxx11::string = std::__cxx11::basic_string<char>]’, declared using local type ‘const main()::<lambda(int, int&)>’, is used but never defined [-fpermissive]
   static inline void execute(const std::string& label, const PolicyType& policy,
                      ^~~~~~~
/tmp/kokkos-alt/core/src/Kokkos_Parallel_Reduce.hpp:1588:22: warning: ‘static void Kokkos::Impl::ParallelReduceAdaptor<PolicyType, FunctorType, ReturnType>::execute(const string&, const PolicyType&, const FunctorType&, ReturnType&) [with PolicyType = Kokkos::RangePolicy<Kokkos::OpenMP>; FunctorType = main()::<lambda(int, int&)>; ReturnType = const int]’ used but never defined
CMakeFiles/example.dir/build.make:62: recipe for target 'CMakeFiles/example.dir/cmake_example.cpp.o' failed
make[2]: *** [CMakeFiles/example.dir/cmake_example.cpp.o] Error 1
CMakeFiles/Makefile2:464: recipe for target 'CMakeFiles/example.dir/all' failed
```
for me.

**Review by dalg24 — APPROVED** [review]

File: core/src/Kokkos_Parallel_Reduce.hpp [hunk 1]
```diff
@@ -1733,10 +1733,15 @@ struct ParallelReduceFence {
...
-inline typename std::enable_if<
-    Kokkos::is_execution_policy<PolicyType>::value>::type
+inline std::enable_if_t<Kokkos::is_execution_policy<PolicyType>::value &&
+                        !(Kokkos::is_view<ReturnType>::value ||
+                          Kokkos::is_reducer<ReturnType>::value)>
 parallel_reduce(const std::string& label, const PolicyType& policy,
                 const FunctorType& functor, ReturnType& return_value) {
+  static_assert(
+      !std::is_const<ReturnType>::value,
+      "A const return type is only allowed for a View or reducer return type!");
```
Comment (dalg24) [inline core/src/Kokkos_Parallel_Reduce.hpp:1743]: Did you check in our Wiki how we actually refer to the reducer argument?  I feel like "return type" might be confusing

Comment (masterleinad) [reply core/src/Kokkos_Parallel_Reduce.hpp]: I changed it to "reduction result type".

**Review by crtrott — APPROVED** [review]

**Comment by masterleinad** [comment]
I needed to also allow pointers to be passed in as `const`.

**Pull request merged**